mailman/submit.py

```diff
@@ -78,67 +64,13 @@ def main(argv):
         sub.save()
         return 0
 
-    with tempfile.TemporaryDirectory() as repo_path:
-        repo = git.Repo.init(repo_path)
-        maildir = Path(MAIL_DIR_ABSPATH)
-        author_args = ["-c", "user.name=Denis", "-c",
-                       "user.email=daemon@mailman.d"]
-        git_am_args = ["git", *author_args, "am", "--keep"]
-        whitespace_errors = []
-
-        def am_cover_letter(keep_empty=True):
-            args = git_am_args
-            if keep_empty:
-                args.append("--empty=keep")
-            repo.git.execute([*args, str(maildir/cover_letter.msg_id)])
-
-        if try_or_false(lambda: am_cover_letter(keep_empty=False),
-                        git.GitCommandError):
-            sub.status = "missing cover letter"
-            sub.save()
-            return 0
-        repo.git.execute(["git", *author_args, "am", "--abort"])
-        if not try_or_false(lambda: am_cover_letter(keep_empty=True),
-                            git.GitCommandError):
-            sub.status = ("missing cover letter and "
-                          "first patch failed to apply")
-            sub.save()
-            return 0
-
-        for i, patch in enumerate(patches):
-            patch_abspath = str(maildir / patch.msg_id)
-
-            # Try and apply and fail if there are whitespace errors
-            def do_git_am(extra_args=[]):
-                repo.git.execute([*git_am_args, *extra_args, patch_abspath]),
-
-            # If this fails, the patch may apply with whitespace errors
-            if try_or_false(lambda: do_git_am(['--whitespace=error-all']),
-                            git.GitCommandError):
-                continue
-
-            repo.git.execute(["git", *author_args, "am", "--abort"])
-
-            # Try again, if we succeed, count this patch as a whitespace error
-            if try_or_false(lambda: do_git_am(), git.GitCommandError):
-                whitespace_errors.append(str(i+1))
-                continue
-
-            # If we still fail, the patch does not apply
-            sub.status = f'patch {i+1} failed to apply'
-            sub.save()
-            return 0
-
-        if whitespace_errors:
-            sub.status = ('whitespace error patch(es) '
-                          f'{",".join(whitespace_errors)}')
-        else:
-            sub.status = 'patchset applies'
-        sub.save()
+    status, repo_path = patchset.check(cover_letter, patches)
+    sub.status = status
+    sub.save()
+
+    patchset.tag_and_push(repo_path, logfile)
 
-        repo.create_tag(logfile)
-        repo.create_remote("origin", REMOTE_URL)
-        repo.git.push("origin", tags=True)
+    repo_path.cleanup()
```

Comment:
Requiring this call to cleanup is really ugly and while my first thought was that we could just sweep it under the rug by using a `with` block like
```py
with repo_path:
    patchset.tag_and_push(repo_path, logfile)
```
it doesn't fix the fundamental problem of needing to remember to handle this thing in some sort of special way because we are exposing the internal implementation details of that module (that it is using a temp dir) to code outside of the module.

Fundamentally this feels like too much guts in of the other code being touched here. I think we should put all code that wants to work with the repo path back into a with block within the module and just make a function that does all the checking, pushing and tagging, and then returns the status so that the code that goes in here is just `sub.status = patchset.check(cover_letter, patches); sub.save()`

I know that in some cases we had said that might not want to tag and push, but I think we could go back add a parameter with a default value to skip tagging and pushing later if we want that.